webui/react/src/components/ColumnsCustomizeModal.tsx

```diff
@@ -0,0 +1,59 @@
+import React, { useEffect, useRef, useState } from 'react';
+
+import { Modal } from 'components/kit/Modal';
+import Transfer from 'components/Transfer';
+import usePrevious from 'shared/hooks/usePrevious';
+import { isEqual } from 'shared/utils/data';
+
+interface Props {
+  columns: string[];
+  defaultVisibleColumns: string[];
+  initialVisibleColumns?: string[];
+  onSave?: (columns: string[]) => void;
+}
+
+const ColumnsCustomizeModalComponent: React.FC<Props> = ({
+  columns,
+  defaultVisibleColumns,
+  initialVisibleColumns,
+  onSave,
+}: Props) => {
+  const columnList = useRef(columns).current; // This is only to prevent rerendering
+  const prevInitVisibleColumns = usePrevious(initialVisibleColumns, defaultVisibleColumns);
+  const [visibleColumns, setVisibleColumns] = useState<string[]>(
+    initialVisibleColumns ?? defaultVisibleColumns,
+  );
+
+  useEffect(() => {
+    // If you travel between pages that both use this hook `visibleColumns` doesn't reinitialize.
+    // That means that when you open this modal on the second page it will have the value of `visibleColumns` from the first page.
+    // This useEffect makes sure that if the value of `initialVisibleColumns` prop changes that
+    // `visibleColumns` will be set to the new value.
+    if (!isEqual(initialVisibleColumns, prevInitVisibleColumns))
+      setVisibleColumns(initialVisibleColumns ?? defaultVisibleColumns);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [initialVisibleColumns]);
```

Comment:
let's just use a key on the component to force re-initialization.

Reply:
I think it evens out because the component will double-render when the initialVisibleColumns is changed with the effect.

Reply:
actually not seeing the behavior described without the `useEffect`, so will just remove it.